tests/cases/conformance/types/thisType/thisTypeInObjectLiterals2.ts

```diff
@@ -0,0 +1,144 @@
+// @declaration: true
+// @noImplicitAny: true
+// @noImplicitThis: true
+// @target: es5
+
+// In methods of an object literal with no contextual type, 'this' has the type
+// of the object literal.
+
+let obj1 = {
+    a: 1,
+    f() {
+        return this.a;
+    },
+    b: "hello",
+    c: {
+        g() {
+            this.g();
+        }
+    },
+    get d() {
+        return this.a;
+    },
+    get e() {
+        return this.b;
+    },
+    set e(value) {
+        this.b = value;
+    }
+};
+
+// In methods of an object literal with a contextual type, 'this' has the
+// contextual type.
```

Comment:
Can you add a similar test based on the RWC case that used optional properties from the contextual type? Something like

```ts
type Point = {
  x: number;
  y: number;
  z?: number;
  moveBy(dx: number, dy: number, dz?: number): void;
}
let p1: Point = {
  x: 10,
  y: 20,
  backward: true,
  moveBy(dx, dy, dz) {
    // 'backward' from object literal is available
    this.x += dx * (this.backward ? -1 : 1);
    this.y += dy;
    // 'z' from contextual type is available
    if (this.z && dz) {
      this.z += dz;
    }
  }
}
```